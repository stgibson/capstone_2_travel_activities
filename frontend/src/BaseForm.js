import React, { useState, useEffect, useCallback } from "react";
import Form from "react-bootstrap/Form";
import Button from "react-bootstrap/Button";

/**
 * Component to create template for forms used in other components
 * @param {Object} param0 
 * @returns JSX code for rendering a form
 */
const BaseForm = ({ title, inputs, btnText, submitCallback }) => {
  const getFormData = useCallback(() => {
    return inputs.reduce((obj, input) => {
      if (input.type === "select") {
        obj[input.name] = input.choices.length ? `${input.choices[0].id}` :
          undefined;
      }
      else {
        obj[input.name] = input.default;
      }
      return obj;
    }, {});
  }, [inputs]);

  // inputs should be array of name, label, type, default, & opt. choices
  const initFormData = getFormData();
  const [formData, setFormData] = useState(initFormData);

  /**
   * Updates form input when user types in input
   * @param {Object} evt 
   */
  const handleChange = evt => {
    const { id, value } = evt.target;
    setFormData(formData => ({ ...formData, [id]: value }));
  };

  /**
   * When user submits form, prevents page from reloading, resets form, and
   * calls submitCallback
   * @param {Object} evt
   */
  const handleSubmit = evt => {
    evt.preventDefault();
    setFormData(initFormData);
    submitCallback(formData);
  };

  // if inputs changes, updated form data
  useEffect(() => {
    const newFormData = getFormData();
    setFormData(newFormData);
  }, [inputs, getFormData]);
  
  return (
    <>
      { title && <h2>{ title }</h2> }
      <Form>
        {
          inputs.map(input => {
            if (input.type === "select") {
              return (
                <Form.Group key={ input.name } controlId={ input.name }>
                  <Form.Label>{ input.label }</Form.Label>
                  <Form.Control as="select" onChange={ handleChange } value={ formData[input.name] }>
                    {
                      input.choices.map(choice => (
                        <option key={ choice.id } value={ choice.id }>
                          { choice.name }
                        </option>
                      ))
                    }
                  </Form.Control>
                </Form.Group>
              )
            }
            return (
              <Form.Group key={ input.name } controlId={ input.name }>
                <Form.Label>{ input.label }</Form.Label>
                <Form.Control
                  type={ input.type }
                  onChange={ handleChange }
                  value={ formData[input.name] }
                />
              </Form.Group>
            )
          })
        }
      </Form>
      <Button variant="primary" onClick={ handleSubmit }>{ btnText }</Button>
    </>
  );
};

export default BaseForm;