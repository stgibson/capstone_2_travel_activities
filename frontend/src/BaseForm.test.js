import React from "react";
import { render } from "@testing-library/react";
import BaseForm from "./BaseForm";

const inputs = [
  { name: "input1", label: "Input 1", type: "text", default: "" },
  { name: "input2", label: "Input 2", type: "text", default: "" }
];
const callback = jest.fn();

it("renders without crashing", () => {
  render(
    <BaseForm
      title="Form"
      inputs={ inputs }
      btnText="Button Text"
      submitCallback={ callback }
    />
  );
});

it("matches snapshot", () => {
  const { asFragment } = render(
    <BaseForm
      title="Form"
      inputs={ inputs }
      btnText="Button Text"
      submitCallback={ callback }
    />
  );

  expect(asFragment()).toMatchSnapshot();
});
