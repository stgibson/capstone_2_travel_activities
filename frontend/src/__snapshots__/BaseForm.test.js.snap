// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <h2>
    Form
  </h2>
  <form
    class=""
  >
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="input1"
      >
        Input 1
      </label>
      <input
        class="form-control"
        id="input1"
        type="text"
        value=""
      />
    </div>
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="input2"
      >
        Input 2
      </label>
      <input
        class="form-control"
        id="input2"
        type="text"
        value=""
      />
    </div>
  </form>
  <button
    class="btn btn-primary"
    type="button"
  >
    Button Text
  </button>
</DocumentFragment>
`;
